nuxeo-ai-model/src/main/java/org/nuxeo/ai/bulk/BulkEnrichmentAction.java

```diff
@@ -0,0 +1,109 @@
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.nuxeo.ai.model.serving.ModelServingService;
+import org.nuxeo.ai.pipes.functions.PropertyUtils;
+import org.nuxeo.ai.pipes.types.BlobTextFromDocument;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.DocumentNotFoundException;
+import org.nuxeo.ecm.core.api.IdRef;
+import org.nuxeo.ecm.core.bulk.action.computation.AbstractBulkComputation;
+import org.nuxeo.ecm.core.bulk.message.BulkStatus;
+import org.nuxeo.lib.stream.computation.ComputationContext;
+import org.nuxeo.lib.stream.computation.Record;
+import org.nuxeo.lib.stream.computation.Topology;
+import org.nuxeo.runtime.api.Framework;
+import org.nuxeo.runtime.stream.StreamProcessorTopology;
+
+/**
+ * A BAF action to add enrichment or suggestions to a number of documents. It does this by sending a sub-set of the
+ * document properties to a stream for downstream processing. The current implementation only works with custom models.
+ */
+public class BulkEnrichmentAction implements StreamProcessorTopology {
+
+    public static final String ACTION_NAME = "bulkEnrich";
+
+    public static final String STREAM_NAME = "streamName";
+
+    @Override
+    public Topology getTopology(Map<String, String> options) {
+
+        String outputStream = options.get(STREAM_NAME);
+        return Topology.builder()
+                       .addComputation(DocEnrichingComputation::new,
+                               Arrays.asList(INPUT_1 + ":" + ACTION_NAME, OUTPUT_1 + ":" + STATUS_STREAM,
+                                       OUTPUT_2 + ":" + outputStream))
+                       .build();
+    }
+
+    public static class DocEnrichingComputation extends AbstractBulkComputation {
+
+        private static final Logger log = LogManager.getLogger(DocEnrichingComputation.class);
+
+        List<Record> outputs = new ArrayList<>();
+
+        public DocEnrichingComputation() {
+            super(ACTION_NAME, 2);
+        }
+
+        @Override
+        protected void compute(CoreSession coreSession, List<String> ids, Map<String, Serializable> map) {
+            ModelServingService modelServingService = Framework.getService(ModelServingService.class);
+            for (String id : ids) {
+                try {
+                    DocumentModel doc = coreSession.getDocument(new IdRef(id));
+                    Set<String> inputs = modelServingService.getInputs(doc);
+                    if (!inputs.isEmpty()) {
+                        BlobTextFromDocument blobTextFromDocument = PropertyUtils.docSerialize(doc, inputs);
+                        if (blobTextFromDocument != null) {
+                            outputs.add(toRecord(blobTextFromDocument.getKey(), blobTextFromDocument));
```

Comment:
how large outputs can be? Do we really need to keep the array on the class level?

Reply:
Yes, that's a bit nasty isn't it.  Its the way other batch computations.  You need the `ComputationContext` to write the record.  The batch size is set by config, so its 50.